src/service/basket.test.js

```diff
@@ -1,58 +1,122 @@
     basket.addProduct(product2);
+
     expect(basket.get().products.length).toEqual(2);
-    expect(basket.get().total).toEqual(1597);
+    expect(basket.get().total).toEqual(product.price + product2.price);
   });
 
-  it('that product can be removed from the basket', () => {
-    basket.addProduct(mockProduct);
+  it('should increment products count when adding the same product name but different size', () => {
+    const product = createProduct();
+    const product2 = createProduct({ name : product.name });
+    
+    basket.addProduct(product);
     basket.addProduct(product2);
-    basket.removeProduct(mockProduct.name);
+
+    expect(basket.get().products.length).toEqual(2);
+    expect(product.quantity).toEqual(1);
+  });
+
+  it('should add the same product and increment its quantity', () => {
+    const product = createProduct();
+
+    basket.addProduct(product);
+    basket.addProduct(product);
+
     expect(basket.get().products.length).toEqual(1);
-    expect(basket.get().total).toEqual(342);
+    expect(product.quantity).toEqual(2);
   });
 
-  it('does basket.applyDiscount for amount work', () => {
+  it('should remove product from the basket and change products price and amount', () => {
+    const product = createProduct();
+    const product2 = createProduct();
+
+    basket.addProduct(product);
     basket.addProduct(product2);
-    basket.applyDiscount(20, false);
-    expect(basket.get().total).toEqual(322);
+    basket.removeProduct(product.name, product.size);
+
+    expect(basket.get().products.length).toEqual(1);
+    expect(basket.get().total).toEqual(product2.price);
+  });
+
+  it('should decrement product quantity when removed the same product', () => {
+    const product = createProduct();
+    
+    basket.addProduct(product);
+    basket.addProduct(product);
+    basket.removeProduct(product.name, product.size);
+
+    expect(product.quantity).toEqual(1);
   });
 
-  it('does basket.applyDiscount for percentage work', () => {
+  it('should decrement basket count when removed the same product but different size', () => {
+    const product = createProduct();
+    const product2 = createProduct({ name : product.name });
+    
+    basket.addProduct(product);
     basket.addProduct(product2);
+    basket.removeProduct(product.name, product.size);
+    
+    expect(basket.get().products.length).toEqual(1)
+  });
+
+  it('should apply a percentage discount on the price of the product', () => {
+    const product = createProduct();
+    product.price = 100;
+
+    basket.addProduct(product);
     basket.applyDiscount(20);
-    expect(basket.get().total).toEqual(273.6);
+    
+    expect(basket.total).toEqual(80);
   });
 
-  it('if basket can be cleared ', () => {
+  it('should apply amount discount for the product price', () => {
+    const product = createProduct();
+ 
+    basket.addProduct(product);
+    basket.applyDiscount(20, false);
+
+    expect(product.price).toEqual(basket.total + 20);
```

Comment:
I think this assertion should change, we should be checking discount applied on basket not product.
This can be tricky in this situation, let me know if you need help